Snake and Ladder Problem.md

```diff
@@ -0,0 +1,121 @@
+# Snake and Ladder Problem 
+
+Problem statement: Given that you have control on the outcome of the dice rolled in a snake and ladder game, find the minimum number of throws required to reach the end of the game. Below is the image of a snake and ladder board on which we have to find the number of throws.
+
+![1571553012293](C:\Users\risha\AppData\Roaming\Typora\typora-user-images\1571553012293.png)
+
+
+
+
+
+## How do we proceed further ?
+
+Here are the broad steps that one can keep in mind:
+
+ *  Convert the problem into a graph problem
+ *  The idea is to consider the given snake and ladder board as a directed
+    graph with number of vertices equal to the number of cells in the board
+ *  Since, all the edges are of "equal weight", it can be solved using breadth first search to find the shortest path
+
+
+
+
+
+
+
+
+
+
+
+# Algorithm Discussion
+
+Following the steps as discussed above, convert the problem into graphs. Each cell on the board will be a node in the graph. The idea is to reach the destination node in the minimum number of steps from the source node. It is nothing but a problem to find the shortest path to reach the destination node. Since, all the edges are of equal weight - we can use `breadth first search` to find the same.
+
+The other data structures that we will require are: A queue, A visited array and a collection/class/structure that will contain two values - Value of the vertex and the number of tries (number of times dice is rolled) to reach the vertex. 
+
+This structure/class/collection will be enqueued. So, let us enqueue the first or the source node in the queue and mark the index in the visited array as 1 (means that node is visited). It is important to note that at any given index, the next move will be `  v+1, v+2, v+3, v+4, v+5 or v+6`, where v is the current index.
+
+Now, we dequeue the head of the queue and enqueue the next six possible nodes. Before enqueuing we have to check if the node is visited or not, if it is visited we need not enqueue (it will already be in the queue) and if it is not visited - we enqueue it and change the value as visited. 

```

Comment:
I think using points to show each step would be more concise

Reply:
Yes true, actually we want this to be conversational in nature, So that people can come up with more optimised solutions. This is not the final solution. People are free to suggest more changes. Some loopholes have been left there for purpose